### ISSUE TITLE
Prefect context manager overwrites global contexts only once

### ISSUE THREAD

**Issue by k24d** [PrefectHQ/prefect#3828]
## Description

The [document](https://docs.prefect.io/core/concepts/execution.html#modifying-context-at-runtime) states we can modify global context keys by using a context manager:

> ### Modifying context at runtime
> Modifying context, even globally set context keys, at specific times is possible using a provided context manager:

However, in the attached script, the context key `foo` is overwritten only once.

## Expected Behavior

Modified context values do not change during a flow run.

### context-test.py
```python
import prefect
from prefect import task, Flow

@task
def print_contexts():
    print("foo = " + prefect.context.foo)

with Flow('test') as flow:
    t1 = print_contexts()
    t2 = print_contexts()
    t1.set_downstream(t2)

with prefect.context(foo='foo'):
    flow.run()
```

### ~/.prefect/config.toml
```
[context]
foo = "global"
```

### Execution result
```
% python context-test.py
[2020-12-14 16:12:46+0900] INFO - prefect.FlowRunner | Beginning Flow run for 'test'
[2020-12-14 16:12:46+0900] INFO - prefect.TaskRunner | Task 'print_contexts': Starting task run...
foo = foo
[2020-12-14 16:12:46+0900] INFO - prefect.TaskRunner | Task 'print_contexts': Finished task run for task with final state: 'Success'
[2020-12-14 16:12:46+0900] INFO - prefect.TaskRunner | Task 'print_contexts': Starting task run...
foo = global  <-- UNEXPECTED!!
[2020-12-14 16:12:46+0900] INFO - prefect.TaskRunner | Task 'print_contexts': Finished task run for task with final state: 'Success'
[2020-12-14 16:12:46+0900] INFO - prefect.FlowRunner | Flow run SUCCESS: all reference tasks succeeded
```

## Environment

```
% prefect diagnostics
{
  "config_overrides": {
    "context": {
      "foo": true
    }
  },
  "env_vars": [],
  "system_information": {
    "platform": "macOS-10.16-x86_64-i386-64bit",
    "prefect_backend": "cloud",
    "prefect_version": "0.13.19",
    "python_version": "3.8.5"
  }
}
```

**Comment by joshmeek** [comment]
@k24d Very weird! I'll look into this 🤔

**Comment by jcrist** [comment]
Thanks for the issue report, this should be fixed by #3924.

```text
$ cat test.py
import prefect
from prefect import task, Flow


@task
def print_contexts():
    print("foo = " + prefect.context.foo)


with Flow('test') as flow:
    t1 = print_contexts()
    t2 = print_contexts()
    t1.set_downstream(t2)


with prefect.context(foo='foo'):
    flow.run()

$ python test.py
[2021-01-05 16:56:28-0600] INFO - prefect.FlowRunner | Beginning Flow run for 'test'
[2021-01-05 16:56:28-0600] INFO - prefect.TaskRunner | Task 'print_contexts': Starting task run...
foo = foo
[2021-01-05 16:56:28-0600] INFO - prefect.TaskRunner | Task 'print_contexts': Finished task run for task with final state: 'Success'
[2021-01-05 16:56:28-0600] INFO - prefect.TaskRunner | Task 'print_contexts': Starting task run...
foo = foo
[2021-01-05 16:56:28-0600] INFO - prefect.TaskRunner | Task 'print_contexts': Finished task run for task with final state: 'Success'
[2021-01-05 16:56:28-0600] INFO - prefect.FlowRunner | Flow run SUCCESS: all reference tasks succeeded
```
